File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/ui/TransactionDetailDialog.java
package org.mifos.mobilewallet.mifospay.history.ui;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.BottomSheetDialog;
import android.support.design.widget.BottomSheetDialogFragment;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;
import org.mifos.mobilewallet.core.data.fineractcn.entity.journal.Account;
import org.mifos.mobilewallet.core.data.fineractcn.entity.journal.JournalEntry;
import org.mifos.mobilewallet.mifospay.R;
import org.mifos.mobilewallet.mifospay.base.BaseActivity;
import org.mifos.mobilewallet.mifospay.data.local.PreferencesHelper;
import org.mifos.mobilewallet.mifospay.history.HistoryContract;
import org.mifos.mobilewallet.mifospay.history.presenter.TransactionDetailPresenter;
import org.mifos.mobilewallet.mifospay.utils.Constants;
import org.mifos.mobilewallet.mifospay.utils.Utils;
import org.mifos.mobilewallet.mifospay.utils.Toaster;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static org.mifos.mobilewallet.mifospay.utils.Constants.CREDIT;
import static org.mifos.mobilewallet.mifospay.utils.Constants.DEBIT;
import static org.mifos.mobilewallet.mifospay.utils.Constants.OTHER;

/**
 * Created by ankur on 06/June/2018
 */

public class TransactionDetailDialog extends BottomSheetDialogFragment implements
        HistoryContract.TransactionDetailView {

    @Inject
    TransactionDetailPresenter mPresenter;
    HistoryContract.TransactionDetailPresenter mTransactionDetailPresenter;

    @Inject
    PreferencesHelper preferencesHelper;

    @BindView(R.id.tv_transaction_id)
    TextView tvTransactionId;

    @BindView(R.id.tv_transaction_date)
    TextView tvTransactionDate;

    @BindView(R.id.tv_receiptId)
    TextView tvReceiptId;

    @BindView(R.id.tv_transaction_status)
    TextView tvTransactionStatus;

    @BindView(R.id.tv_transaction_amount)
    TextView tvTransactionAmount;

    @BindView(R.id.rl_from_to)
    RelativeLayout rlFromTo;

    @BindView(R.id.iv_fromImage)
    ImageView ivFromImage;

    @BindView(R.id.tv_fromClientName)
    TextView tvFromClientName;

    @BindView(R.id.tv_fromAccountNo)
    TextView tvFromAccountNo;

    @BindView(R.id.iv_toImage)
    ImageView ivToImage;

    @BindView(R.id.tv_toClientName)
    TextView tvToClientName;

    @BindView(R.id.tv_toAccountNo)
    TextView tvToAccountNo;

    @BindView(R.id.v_rule2)
    View vRule2;

    @BindView(R.id.tv_viewReceipt)
    TextView tvViewReceipt;
    @BindView(R.id.ll_from)
    LinearLayout mLlFrom;
    @BindView(R.id.ll_to)
    LinearLayout mLlTo;
    @BindView(R.id.pb_transaction_detail)
    ProgressBar progressBar;
    @BindView(R.id.ll_main)
    LinearLayout mLlMain;

    private BottomSheetBehavior mBottomSheetBehavior;
    private JournalEntry transaction;
    private ArrayList<JournalEntry> transactions;
    private String transactionType;

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        BottomSheetDialog dialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);

        View view = View.inflate(getContext(), R.layout.dialog_transaction_detail, null);

        dialog.setContentView(view);
        mBottomSheetBehavior = BottomSheetBehavior.from((View) view.getParent());

        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        ButterKnife.bind(this, view);

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            transaction = bundle.getParcelable(Constants.TRANSACTION);
            transactions = bundle.getParcelableArrayList(Constants.TRANSACTIONS);
            if (transaction == null) {
                return dialog;
            }
        }
        mPresenter.attachView(this);
        showDetails(transaction);
        return dialog;
    }

    @Override
    public void onStart() {
        super.onStart();
        mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
    }

    @Override
    public void setPresenter(HistoryContract.TransactionDetailPresenter presenter) {
        mTransactionDetailPresenter = presenter;
    }

    @OnClick(R.id.tv_viewReceipt)
    public void viewReceipt() {
//        Intent intent = new Intent(getActivity(), ReceiptActivity.class);
//        intent.setData(Uri.parse(Constants.RECEIPT_DOMAIN + transaction.getTransactionId()));
//        startActivity(intent);
    }

    @OnClick(R.id.rl_from_to)
    public void showSpecificTransactions() {

        Intent intent = new Intent(getActivity(), SpecificTransactionsActivity.class);
        intent.putParcelableArrayListExtra(Constants.TRANSACTIONS, transactions);
        if (transactionType == DEBIT) {
            intent.putExtra(Constants.ACCOUNT_NUMBER, tvFromAccountNo.getText().toString());
            intent.putExtra(Constants.OTHER_CUSTOMER_NAME, tvFromClientName.getText().toString());
        } else if (transactionType == CREDIT) {
            intent.putExtra(Constants.ACCOUNT_NUMBER, tvToAccountNo.getText().toString());
            intent.putExtra(Constants.OTHER_CUSTOMER_NAME, tvToClientName.getText().toString());
        }
        startActivity(intent);

    }

    private void showDetails(JournalEntry transaction) {

        try {
            tvTransactionId.setText(Constants.TRANSACTION_ID + ": " +
                    transaction.getTransactionIdentifier());
            tvTransactionDate.setText(Constants.DATE + ": " + transaction.getTransactionDate());
            List<Account> creditors = transaction.getCreditors();
            List<Account> debtors = transaction.getDebtors();
            String creditAccountIdentifier = creditors.get(0).getAccountNumber();
            String debitAccountIdentifier = debtors.get(0).getAccountNumber();

            tvFromAccountNo.setText(creditAccountIdentifier);
            tvToAccountNo.setText(debitAccountIdentifier);

            Double amount = Double.valueOf(creditors.get(0).getAmount());
            transactionType = OTHER;
            if (preferencesHelper.getCustomerDepositAccountIdentifier()
                    .equals(creditAccountIdentifier)) {
                transactionType = CREDIT;
                /**
                 * Fetching the details of the debtor
                 */
                mTransactionDetailPresenter.fetchAccountDetail(debitAccountIdentifier);
                tvFromClientName.setText(preferencesHelper.getCustomerName());
            } else if (preferencesHelper.getCustomerDepositAccountIdentifier()
                    .equals(debitAccountIdentifier)) {
                transactionType = DEBIT;
                /**
                 * Fetching the details of the creditor
                 */
                mTransactionDetailPresenter.fetchAccountDetail(creditAccountIdentifier);
                tvToClientName.setText(preferencesHelper.getCustomerName());
            }

            tvTransactionAmount.setText(Utils.getFormattedAccountBalance(
                    amount, preferencesHelper.getCurrencySign()));

//        if (transaction.getTransactionType() != null) {
//            tvReceiptId.setVisibility(View.VISIBLE);
//            tvReceiptId.setText(Constants.RECEIPT_ID + ": " + transaction.getReceiptId());
//        }

            switch (transactionType) {
                case DEBIT:
                    tvTransactionStatus.setText(DEBIT);
                    tvTransactionAmount.setTextColor(Color.RED);
                    break;
                case CREDIT:
                    tvTransactionStatus.setText(CREDIT);
                    tvTransactionAmount.setTextColor(Color.parseColor("#009688"));
                    break;
                case OTHER:
                    tvTransactionStatus.setText(OTHER);
                    tvTransactionAmount.setTextColor(Color.YELLOW);
                    break;
            }
        } catch (NullPointerException e) {
            showToast(getString(R.string.unexpected_error_subtitle));
            dismiss();
        }
    }


    @Override
    public void showCustomerName(String customerName) {
        rlFromTo.setVisibility(View.VISIBLE);
        vRule2.setVisibility(View.VISIBLE);
        if (transactionType.equals(DEBIT)) {
            tvFromClientName.setText(customerName);
        } else if (transactionType.equals(CREDIT)) {
            tvToClientName.setText(customerName);
        }
    }

    private void showToast(String message) {
        Toaster.showToast(getActivity(), message);
    }

    @Override
    public void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showError(String message) {
        showToast(message);
        dismiss();
    }

}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/ui/TransactionDetailDialog.kt
package org.mifos.mobilewallet.mifospay.history.ui

import android.app.Dialog
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.Unbinder
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobilewallet.core.data.fineract.entity.accounts.savings.TransferDetail
import org.mifos.mobilewallet.core.domain.model.Transaction
import org.mifos.mobilewallet.core.domain.model.TransactionType
import org.mifos.mobilewallet.mifospay.R
import org.mifos.mobilewallet.mifospay.history.HistoryContract
import org.mifos.mobilewallet.mifospay.history.HistoryContract.TransactionDetailView
import org.mifos.mobilewallet.mifospay.history.presenter.TransactionDetailPresenter
import org.mifos.mobilewallet.mifospay.receipt.ui.ReceiptActivity
import org.mifos.mobilewallet.mifospay.utils.Constants
import org.mifos.mobilewallet.mifospay.utils.Toaster
import org.mifos.mobilewallet.mifospay.utils.Utils.getFormattedAccountBalance
import javax.inject.Inject

/**
 * Created by ankur on 06/June/2018
 */
@AndroidEntryPoint
class TransactionDetailDialog : BottomSheetDialogFragment(), TransactionDetailView {
    @JvmField
    @Inject
    var mPresenter: TransactionDetailPresenter? = null
    var mTransactionDetailPresenter: HistoryContract.TransactionDetailPresenter? = null

    @JvmField
    @BindView(R.id.tv_transaction_id)
    var tvTransactionId: TextView? = null

    @JvmField
    @BindView(R.id.tv_transaction_date)
    var tvTransactionDate: TextView? = null

    @JvmField
    @BindView(R.id.tv_receiptId)
    var tvReceiptId: TextView? = null

    @JvmField
    @BindView(R.id.tv_transaction_status)
    var tvTransactionStatus: TextView? = null

    @JvmField
    @BindView(R.id.tv_transaction_amount)
    var tvTransactionAmount: TextView? = null

    @JvmField
    @BindView(R.id.rl_from_to)
    var rlFromTo: RelativeLayout? = null

    @JvmField
    @BindView(R.id.iv_fromImage)
    var ivFromImage: ImageView? = null

    @JvmField
    @BindView(R.id.tv_fromClientName)
    var tvFromClientName: TextView? = null

    @JvmField
    @BindView(R.id.tv_fromAccountNo)
    var tvFromAccountNo: TextView? = null

    @JvmField
    @BindView(R.id.iv_toImage)
    var ivToImage: ImageView? = null

    @JvmField
    @BindView(R.id.tv_toClientName)
    var tvToClientName: TextView? = null

    @JvmField
    @BindView(R.id.tv_toAccountNo)
    var tvToAccountNo: TextView? = null

    @JvmField
    @BindView(R.id.v_rule2)
    var vRule2: View? = null

    @JvmField
    @BindView(R.id.tv_viewReceipt)
    var tvViewReceipt: TextView? = null

    @JvmField
    @BindView(R.id.ll_from)
    var mLlFrom: LinearLayout? = null

    @JvmField
    @BindView(R.id.ll_to)
    var mLlTo: LinearLayout? = null

    @JvmField
    @BindView(R.id.pb_transaction_detail)
    var progressBar: ProgressBar? = null
    var unbinder: Unbinder? = null

    @JvmField
    @BindView(R.id.ll_main)
    var mLlMain: LinearLayout? = null
    private var mBottomSheetBehavior: BottomSheetBehavior<*>? = null
    private var transaction: Transaction? = null
    private var accountNo: String? = null
    private var transactions: ArrayList<Transaction>? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
        val view = View.inflate(context, R.layout.dialog_transaction_detail, null)
        dialog.setContentView(view)
        mBottomSheetBehavior = BottomSheetBehavior.from(view.parent as View)
        ButterKnife.bind(this, view)
        val bundle = this.arguments
        if (bundle != null) {
            transaction = bundle.getParcelable(Constants.TRANSACTION)
            accountNo = bundle.getString(Constants.ACCOUNT_NUMBER)
            transactions = bundle.getParcelableArrayList(Constants.TRANSACTIONS)
            if (transaction == null) {
                return dialog
            }
        }
        mPresenter!!.attachView(this)
        tvTransactionId!!.text =
            Constants.TRANSACTION_ID + ": " + transaction!!.transactionId
        tvTransactionDate!!.text =
            Constants.DATE + ": " + transaction!!.date
        tvTransactionAmount!!.text =
            getFormattedAccountBalance(
                transaction!!.amount, transaction!!.currency.code
            )
        mPresenter!!.getTransferDetail(transaction!!.transferId)
        if (transaction!!.receiptId != null) {
            tvReceiptId!!.visibility = View.VISIBLE
            tvReceiptId!!.text =
                Constants.RECEIPT_ID + ": " + transaction!!.receiptId
        }
        when (transaction!!.transactionType) {
            TransactionType.DEBIT -> {
                tvTransactionStatus!!.text = Constants.DEBIT
                tvTransactionAmount!!.setTextColor(Color.RED)
            }

            TransactionType.CREDIT -> {
                tvTransactionStatus!!.text = Constants.CREDIT
                tvTransactionAmount!!.setTextColor(Color.parseColor("#009688"))
            }

            TransactionType.OTHER -> {
                tvTransactionStatus!!.text = Constants.OTHER
                tvTransactionAmount!!.setTextColor(Color.YELLOW)
            }
        }
        return dialog
    }

    override fun onStart() {
        super.onStart()
        mBottomSheetBehavior!!.state = BottomSheetBehavior.STATE_EXPANDED
    }

    @OnClick(R.id.tv_viewReceipt)
    fun viewReceipt() {
        val intent = Intent(activity, ReceiptActivity::class.java)
        intent.data =
            Uri.parse(Constants.RECEIPT_DOMAIN + transaction!!.transactionId)
        startActivity(intent)
    }

    @OnClick(R.id.ll_from)
    fun onFromViewClicked() {
        val intent = Intent(activity, SpecificTransactionsActivity::class.java)
        intent.putParcelableArrayListExtra(Constants.TRANSACTIONS, transactions)
        intent.putExtra(Constants.ACCOUNT_NUMBER, tvFromAccountNo!!.text.toString())
        startActivity(intent)
    }

    @OnClick(R.id.ll_to)
    fun onToViewClicked() {
        val intent = Intent(activity, SpecificTransactionsActivity::class.java)
        intent.putParcelableArrayListExtra(Constants.TRANSACTIONS, transactions)
        intent.putExtra(Constants.ACCOUNT_NUMBER, tvToAccountNo!!.text.toString())
        startActivity(intent)
    }

    override fun showTransferDetail(transferDetail: TransferDetail?) {
        if (transaction!!.transferId != 0L) {
            rlFromTo!!.visibility = View.VISIBLE
            vRule2!!.visibility = View.VISIBLE
            tvFromClientName!!.text = transferDetail?.fromClient?.displayName
            tvFromAccountNo!!.text = transferDetail?.fromAccount?.accountNo
            tvToClientName!!.text = transferDetail?.toClient?.displayName
            tvToAccountNo!!.text = transferDetail?.toAccount?.accountNo
        }
    }

    override fun showProgressBar() {
        progressBar!!.visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        progressBar!!.visibility = View.GONE
    }

    override fun showToast(message: String?) {
        Toaster.showToast(activity, message)
    }

    override fun setPresenter(presenter: HistoryContract.TransactionDetailPresenter?) {
        mTransactionDetailPresenter = presenter
    }

}